### PR TITLE
feat(POD-011): C-Render Markdown emitter (US-2)

### PR DESCRIPTION
**Pull request by invaderDMG**
## Summary
- POD-011 — pure-function `render(segments, transcripts, fmt) -> str` in `src/podcast_script/render.py`, replacing the POD-008-era Protocol-only stub.
- Drives the SRS §1.6 output shape end-to-end: English `> [<ts> — music starts]` / `> [<ts> — music ends]` blockquotes around music regions, ` `<ts>`  text ` for speech lines, time-ordered, `MM:SS` or `HH:MM:SS` per the Pipeline-supplied `fmt`.
- Closes POD-011 (US-2 Must, sprint SP-2 of `PROJECT_PLAN.md`).

## Acceptance criteria satisfied
- [x] AC-US-2.1 — `tests/test_render.py::test_music_marker_exact_shape_AC_US_2_1`, `::test_multiple_music_regions_each_get_a_pair_AC_US_2_1`
- [x] AC-US-2.2 — `tests/test_render.py::test_markers_are_english_regardless_of_speech_language_AC_US_2_2`
- [x] AC-US-2.3 / NFR-3 — `tests/test_render.py::test_emitted_lines_are_in_non_decreasing_time_order_AC_US_2_3`, `::test_speech_inside_music_window_still_orders_correctly_AC_US_2_3`
- [x] AC-US-2.4 — `tests/test_render.py::test_mm_ss_format_for_short_inputs_AC_US_2_4`, `::test_hh_mm_ss_format_for_long_inputs_AC_US_2_4`
- [x] AC-US-2.5 / NFR-4 — `tests/test_render.py::test_drops_noise_and_silence_segments_AC_US_2_5`, `::test_speech_only_input_emits_no_blockquote_AC_US_2_5`

## Edge cases covered
- [x] EC-1 (music bed under speech) — `::test_speech_inside_music_window_still_orders_correctly_AC_US_2_3` exercises the pathological case where a transcript's `start` lies inside an open music region; the priority-tagged sort still produces non-decreasing leading timestamps.

## Architectural notes
- ADR-0007 flat src layout — `src/podcast_script/render.py` already in place from POD-008; this PR fills in the impl.
- ADR-0017 Tier 1 — `tests/test_render.py` is pure-function unit tests with no fakes and no I/O.
- ADR-0006 — `render` is pure (no `sys.exit`, no exceptions); the Pipeline owns I/O failure modes.
- Renderer takes a Pipeline-chosen `fmt: Literal["MM:SS", "HH:MM:SS"]`; `Pipeline._choose_fmt` already feeds the right value (POD-008).
- Tie-breaking on equal timestamps: `music_end` < `music_start` < `speech` so back-to-back regions and EC-1 overlaps render correctly without violating NFR-3.

## Documentation updated
- [x] Module-level docstring in `src/podcast_script/render.py` traces the contract surface back to AC-US-2.1..2.5 and NFR-3 / NFR-4.
- [ ] CHANGELOG / README — N/A at SP-2; both files are scheduled (POD-036 in SP-7, POD-038 in SP-8). Output-format shape becomes user-facing only once the real backends land in SP-3 / SP-4.

## Test plan
- [x] `uv run pytest -q` — 111 tests green (was 102 on `main`; +9 in `tests/test_render.py`)
- [x] `uv run ruff check .` clean
- [x] `uv run ruff format --check .` clean
- [x] `uv run mypy --strict src tests` clean
- [x] No regressions in existing `tests/test_pipeline.py` — Pipeline injects its own `_stub_render`, decoupled from this module's impl
- [ ] CI matrix (Ubuntu + macOS-14) green — see check status below

## Definition of Done
- [x] Code on `feat/POD-011-render-markdown`, squash-merge target per `PROJECT_PLAN.md` §1.2.
- [x] Tier 1 unit tests added (ADR-0017).
- [x] All AC-US-2.1..2.5 verified in test code.
- [ ] CI matrix green (Ubuntu + macOS) — pending.
- [x] Coverage maintained: render module is pure logic, fully exercised by the 9 new tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/podcast_script/render.py [hunk 1]
```diff
@@ -1,25 +1,59 @@
...
+
+_EventKind = Literal["music_start", "music_end", "speech"]
+
+
+@dataclass(frozen=True, slots=True)
+class _Event:
+    """One renderable line plus its sort key."""
+
+    time_s: float
+    priority: int
```
Comment (invaderDMG) [inline src/podcast_script/render.py]: **[suggestion]** [solid] — `priority` and `kind` are redundant

`priority` is fully determined by `kind` (`music_end` ↔ 0, `music_start` ↔ 1, `speech` ↔ 2) — there's no scenario where they could legitimately diverge. The two-field shape leaves room for a future bug where a caller sets one but not the other.

Cited rule: SOLID/SRP — one source of truth per concept; no architectural rule mandates the current split.

Proposed fix (optional): drop the `priority` field and resolve it from `kind` at sort-key construction:

```suggestion
_PRIO_BY_KIND: dict[_EventKind, int] = {
    "music_end": 0,
    "music_start": 1,
    "speech": 2,
}


@dataclass(frozen=True, slots=True)
class _Event:
    """One renderable line plus its sort key."""

    time_s: float
    kind: _EventKind
    line: str
```

Then in `render()`: `events.sort(key=lambda e: (e.time_s, _PRIO_BY_KIND[e.kind]))`.

Feel free to ignore — the current shape is also fine and arguably clearer at the call site.

File: src/podcast_script/render.py [hunk 2]
```diff
@@ -32,3 +66,92 @@ def __call__(
...
+    §1.6 example shape.
+    """
+    out: list[str] = [events[0].line]
+    for prev, curr in pairwise(events):
+        if prev.kind == "music_start" and curr.kind == "music_end":
+            out.append(curr.line)
+        else:
+            out.append("")
+            out.append(curr.line)
+    return "\n".join(out) + "\n"
```
Comment (invaderDMG) [inline src/podcast_script/render.py:146]: **[praise]** [solid] — clean layout helper

`_join_with_block_separators` factors the SRS §1.6 layout rule (tight `music_start → music_end` pair, blank line everywhere else) into one obvious branch. The example shape from the spec falls out without any special-casing of edge transitions. 👍

File: tests/test_render.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    blockquote lines at all — the renderer never invents markers.
+    """
+    segments = [Segment(0.0, 10.0, "speech")]
+    transcripts = [TranscribedSegment(start=0.0, end=10.0, text="just talking")]
+
+    out = render(segments, transcripts, "MM:SS")
+
+    assert "just talking" in out
+    assert not re.search(r"^> \[", out, flags=re.MULTILINE)
+    assert "music" not in out
```
Comment (invaderDMG) [inline tests/test_render.py:106]: **[suggestion]** [ac-coverage] — add a fully-empty-input test

`render([], [], "MM:SS")` is handled by the early-return at `render.py:81-82`, but no test exercises that branch — it's currently a coverage gap (and the PR body claims `render` is fully exercised by the 9 new tests).

Cited rule: PR-body claim "Coverage maintained: render module is pure logic, fully exercised by the 9 new tests" + ADR-0017 Tier 1 (pure-function unit coverage on the renderer).

Proposed fix — append a 10th test:

```suggestion
def test_speech_only_input_emits_no_blockquote_AC_US_2_5() -> None:
    """AC-US-2.5 corollary: with no music regions, the output has no
    blockquote lines at all — the renderer never invents markers.
    """
    segments = [Segment(0.0, 10.0, "speech")]
    transcripts = [TranscribedSegment(start=0.0, end=10.0, text="just talking")]

    out = render(segments, transcripts, "MM:SS")

    assert "just talking" in out
    assert not re.search(r"^> \[", out, flags=re.MULTILINE)
    assert "music" not in out


def test_empty_input_returns_empty_string() -> None:
    """Edge: with no segments and no transcripts, render returns the
    empty string (no trailing newline)."""
    assert render([], [], "MM:SS") == ""
    assert render([], [], "HH:MM:SS") == ""
```

Minor — feel free to ignore.

File: tests/test_render.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        Segment(90.0, 100.0, "music"),
+    ]
+    transcripts = [
+        TranscribedSegment(start=5.0, end=60.0, text="middle"),
+        TranscribedSegment(start=75.0, end=90.0, text="late"),
+    ]
+
+    out = render(segments, transcripts, "MM:SS")
+
+    assert out.count("music starts") == 3
```
Comment (invaderDMG) [inline tests/test_render.py]: **[suggestion]** [dry] — substring `count` is brittle

`out.count("music starts")` and `out.count("music ends")` count substring occurrences across the whole rendered string, including any speech text that happens to contain the phrase. The current test uses innocuous speech text (`"middle"`, `"late"`) so no current bug, but a future fixture change could break the assertion silently.

Cited rule: AC-US-2.1 verifies marker structure, not text-anywhere — the assertion should reflect that.

Proposed fix — count blockquote lines, not substrings:

```suggestion
    starts = sum(1 for line in out.splitlines() if line.startswith("> [") and "music starts" in line)
    ends = sum(1 for line in out.splitlines() if line.startswith("> [") and "music ends" in line)
    assert starts == 3
    assert ends == 3
```

Applies symmetrically to the same pattern in `test_music_marker_exact_shape_AC_US_2_1` (lines 113-114) if you want to keep it consistent.

**Review by invaderDMG — COMMENTED** [review]
## Review summary

**Would-be verdict:** **approve** (recorded as `event=COMMENT` because GitHub blocks `APPROVE` on a self-authored PR — the skill's review-independence rule was overridden by user request for this session).

**Findings:**
- 0 blocking
- 0 should-fix
- 3 suggestions
- 3 praises (one inline below; two in this summary)

**Quality gate (local, on ff3663a):**
- `uv run pytest -q` → 111 passed (0.34 s); re-run 111 passed (0.26 s) — no flakes, comfortably under the ADR-0017 Tier 1 ≤ 1 s target.
- `uv run ruff check .` → clean.
- `uv run ruff format --check .` → 21 files already formatted.
- `uv run mypy --strict` → no issues, 21 source files.
- CI rollup on the PR matches: `ubuntu-latest` ✅, `macos-14` ✅.

**Top-level notes:**
- **Scope clean.** Diff is exactly `src/podcast_script/render.py` and the new `tests/test_render.py`; no scope creep. POD-011 maps to PROJECT_PLAN.md §SP-2 / §3 (2 sub-pt, US-2 Must); branch matches the agreed naming scheme (§1.2).
- **AC coverage complete.** Every claimed AC (AC-US-2.1..2.5) has at least one named test (`test_*_AC_US_2_N`) that exercises the public `render()` and asserts the SRS §1.6 contract — exact em-dash form, English-only markers, non-decreasing timestamps, no `MM:SS`/`HH:MM:SS` cross-leak, no annotation lines for `noise`/`silence`. EC-1 (music-bed-under-speech) is exercised structurally by the priority-tagged tie-breaking.
- **Architecture respected.** Pure function (ADR-0006: no `sys.exit`, no logging, no I/O) ✓; flat-`src` layout (ADR-0007) ✓; Tier 1 unit test (ADR-0017: pure-function, no fakes, no heavy deps) ✓.
- **TDD discipline — process note (suggestion only).** Two of the cycles (AC-US-2.5 and AC-US-2.3) follow textbook red→green→refactor. Three more (AC-US-2.1, 2.2, 2.4) are property-pinning tests added after the impl already happened to satisfy them — the commit-message bodies are explicit about that, which I appreciate. Worth flagging for the next cycle: when an AC is structurally guaranteed by the design, write the test first to lock the design choice, even if it goes green on first run.
- **Praise (inline + below).** The event-stream-with-priority-tags design is the right shape for the EC-1 ordering edge: explicit, named tie-breaking constants are far cleaner than any merge-walk alternative I considered while reviewing. The module-level docstring's AC traceability is best-in-class for this codebase — every contract clause is bound to an SRS ID, which is exactly what makes the next reviewer's job easy.

**Suggestions (none blocking — all `feel free to ignore`):**
1. `src/podcast_script/render.py` line 53 — collapse the redundant `priority` field (always derivable from `kind`).
2. `tests/test_render.py` near line 96 — add a fully-empty input test for the early-return branch in `render()`.
3. `tests/test_render.py` line 278 — the substring `out.count(...)` is brittle if a transcript ever contained the marker phrase.

Good to merge as soon as you're happy. Re-review not needed for the suggestions.

---

_Posted by `/pr-review` with the review-independence rule overridden by user request. Dismiss with: `gh api -X PUT /repos/invaderDMG/podcast-script/pulls/10/reviews/<id>/dismissals -f message="reviewed in error"`._

**Comment by invaderDMG** [comment]
Addressed all three review suggestions:

1. **`src/podcast_script/render.py` — collapse `_Event.priority`** (commit `bc0dd05`)
   Dropped the redundant field; introduced `_PRIO_BY_KIND` lookup, derived at sort-key construction. SOLID/SRP — single source of truth, removes diverging-state risk. 9 existing tests pass unchanged.

2. **`tests/test_render.py` — empty-input coverage** (commit `ea5e993`)
   Added `test_empty_input_returns_empty_string` (segments=[], transcripts=[] both fmts) and `test_only_noise_and_silence_returns_empty_string_AC_US_2_5` (the renderer never invents content from a no-music / no-speech input). Closes the early-return coverage gap.

3. **`tests/test_render.py` — harden marker counts** (commit `ea5e993`)
   Replaced both `out.count("music starts"/"music ends")` substring assertions with a `_count_marker_lines(out, marker)` helper that counts only blockquote lines (`startswith("> [")`). Applied symmetrically in `test_music_marker_exact_shape_AC_US_2_1` and `test_multiple_music_regions_each_get_a_pair_AC_US_2_1`. AC-US-2.1 invariant is structural, not text-anywhere.

Suite: 111 → 113 green; ruff + mypy --strict clean.

Process note from the review acknowledged — for the next task I'll write the AC-pinning tests *before* the impl even when the design happens to satisfy them on first run.

File: src/podcast_script/render.py [hunk 5]
```diff
@@ -1,25 +1,61 @@
...
 from .backends.base import TranscribedSegment
 from .segment import Segment
 
 TimestampFormat = Literal["MM:SS", "HH:MM:SS"]
 
+_EventKind = Literal["music_start", "music_end", "speech"]
+
+# Tie-breakers when two events share a timestamp (SRS §1.6 ordering):
+# music ends first (closes the open marker), then music starts (opens
+# the next), then speech text (appears at or after any markers active
```
Comment (invaderDMG) [inline src/podcast_script/render.py:41]: **[praise]** [solid] — clean follow-through on the priority/kind collapse

`_PRIO_BY_KIND: dict[_EventKind, int]` puts the tie-break ordering in one obvious lookup, with the `_EventKind` Literal as the dict's key type so a future kind addition that forgot to update this table would surface at `mypy --strict` time rather than silently at runtime. Stronger guarantee than the previous module-level `_PRIO_*` constants. 👍

**Review by invaderDMG — COMMENTED** [review]
## Re-review summary (head `ea5e993`)

**Would-be verdict:** **approve** (recorded as `event=COMMENT` because GitHub blocks `APPROVE` on a self-authored PR — review-independence override carries from the previous turn in this session).

**Findings on the new commits:**
- 0 blocking
- 0 should-fix
- 0 new suggestions
- 1 praise (inline below)

**All three prior suggestions addressed:**
1. ✅ `bc0dd05` — `_Event.priority` field collapsed; sort key now derives via `_PRIO_BY_KIND[e.kind]` lookup. Single source of truth restored. The dict's `dict[_EventKind, int]` typing means a missing kind would surface at type-check time, not silently at runtime — a small upgrade over the previous untyped `_PRIO_*` integers.
2. ✅ `ea5e993` — added `test_empty_input_returns_empty_string` covering both fmt branches, plus a bonus `test_only_noise_and_silence_returns_empty_string_AC_US_2_5` that exercises the noise/silence-drop + early-return composition. Pins the `""` (no trailing newline) contract for empty output.
3. ✅ `ea5e993` — `_count_marker_lines` helper anchors marker counts to `> [` blockquote-line prefixes; `out.count(...)` substring counts replaced in both AC-US-2.1 tests. Helper docstring captures the rationale ("AC-US-2.1 invariant is about marker *structure*, not text-anywhere") — exactly the framing I would have written.

**Quality gate (local, on `ea5e993`):**
- `uv run pytest -q` → **113 passed** (0.33 s); re-run **113 passed** (0.26 s) — still under the ADR-0017 Tier 1 ≤ 1 s target with the +2 new tests.
- `uv run ruff check .` → clean.
- `uv run ruff format --check .` → 21 files already formatted.
- `uv run mypy --strict` → no issues, 21 source files.
- CI rollup: `ubuntu-latest` ✅, `macos-14` ✅ on `ea5e993`.

**No regressions.** The refactor preserves all behavior (the prior 9 tests still pass without modification per `bc0dd05`'s commit body, and the post-refactor 113-test suite is green here). No new attack surface, no new deps, scope still strictly the two files claimed in the original PR body.

**Process note (no action needed):** the two follow-up commits use `fix(POD-011): address review — …` Conventional-Commits prefix. Defensible (treating reviewer-flagged improvements as defects-of-review), though `refactor` for the priority collapse and `test` for the test-only commit would also have been correct. Mild quibble, not a finding.

**Ready to merge** from my POV. The would-be verdict on this re-review is again `approve`; treat the prior review's open suggestions as resolved and proceed when you're ready.

---

_Posted by `/pr-review` (re-review) with the review-independence rule overridden by user request. Dismiss with: `gh api -X PUT /repos/invaderDMG/podcast-script/pulls/10/reviews/<id>/dismissals -f message="reviewed in error"`._
